_stbt/keyboard.py

```diff
@@ -88,70 +82,482 @@ class Keyboard(object):
+            remote control, for example "KEY_RIGHT".
 
-    # pylint:disable=fixme
-    # TODO: case sensitive keyboards
-    #   Caps lock can be supported with a graph like this:
-    #       A CAPSLOCK_OFF KEY_LEFT
-    #       CAPSLOCK_OFF CAPSLOCK_ON KEY_OK
-    #       CAPSLOCK_ON a KEY_RIGHT
-    #       a q KEY_UP
-    #       <etc>
-    #   (Other mode changes like ABC -> 123!@# can be supported in the same
-    #   way.)
-    #
-    #   I don't know how best to support SHIFT (temporary mode change):
-    #   - In shifted state KEY_OK will go from "A" to "a". We'd want to use
-    #     press_and_wait before we check the new state. But...
-    #   - In non-shifted state KEY_OK just enters the letter; the screen doesn't
-    #     change otherwise. Currently we don't use press_and_wait here because
-    #     typically the text-box where the text appears is masked out (because
-    #     some UIs have a blinking cursor there) and there is no change anywhere
-    #     else on the screen.
-    #
-    # TODO: Check that KEY_OK adds to the search text? With a property on the
-    #   page object? OCR might be too unreliable for incomplete words. We don't
-    #   use press_and_wait because the text-box might be masked out (some UIs
-    #   have a blinking cursor there).
+        :param str mode: Optional keyboard mode that applies to both ``source``
+            and ``target``. For example, the two following calls are the same::
+
+                add_transition("c", " ", "KEY_DOWN", mode="lowercase")
+
+                add_transition({"name": "c", "mode": "lowercase"},
+                               {"name": " ", "mode": "lowercase"},
+                               "KEY_DOWN")
+
+        :param bool symmetrical: By default, if the keypress is "KEY_LEFT",
+            "KEY_RIGHT", "KEY_UP", or "KEY_DOWN", this will automatically add
+            the opposite transition. For example, if you call
+            ``add_transition("a", "b", "KEY_RIGHT")`` this will also add the
+            transition ``("b", "a", "KEY_LEFT)"``. Set this parameter to False
+            to disable this behaviour. This parameter has no effect if
+            ``keypress`` is not one of the 4 directional keys.
+
+        :raises: `ValueError` if the ``source`` or ``target`` keys do not exist
+            in the model, or if they can't be identified unambiguously.
+        """
+        source = self._find_key(source, mode)
+        target = self._find_key(target, mode)
+        self._add_edge(source, target, keypress)
+        if symmetrical and keypress in SYMMETRICAL_KEYS:
+            self._add_edge(target, source, SYMMETRICAL_KEYS[keypress])
+
+    def _add_edge(self, source, target, key):
+        # type: (Key, Key, str) -> None
+        self.G.add_edge(source, target, key=key)
+        _add_weight(self.G, source, key)
+        self.G_ = None
+
+    def add_edgelist(self, edgelist, mode=None, symmetrical=True):
```

Comment:
I guess it's for backwards compatibility / easing the upgrade. In practice I suppose you'd use it for defining transitions, not the keys themselves, otherwise you wouldn't be able to identify keys by region. Using OCR won't be fast/reliable enough for identifying keys by name.